config/src/lib.rs

```diff
@@ -1,11 +1,296 @@
+use std::fmt::Error;
+
+use clap::{App, Arg, ArgMatches};
+
+// Include all "stolen" ripgrep code in this module
+mod ripgrep_config;
+
+/// This trait defines the behavior that all configuration classes need to
+/// provide in order for the clap matcher to be generated from the config object
+pub trait ConfigDescription {
+    fn get_config(&self) -> Configuration;
+}
+
+/// This struct describes some properties that can be set for an application as well
+/// as the list of options that the program can understand
+/// These values (apart from the list of options) are only used to generate the
+/// console help message
+pub struct Configuration {
+    /// the name of the application
+    pub name: &'static str,
+    /// version of the application
+    pub version: &'static str,
+    /// a brief description of what the application does
+    pub about: &'static str,
+    /// the list of all possible command line options
+    pub options: Vec<ConfigOption>,
+}
+
+/// Represents an individual config option that the program can interpret
+pub struct ConfigOption {
+    // the name of the option (without leading --)
+    pub name: &'static str,
+    // default value to use for the option if it is not provided
+    pub default: &'static str,
+    // whether this option has to be provided
+    pub required: bool,
+    // if true the option takes a value as argument, if false
+    // the option is a present/missing flag
+    pub takes_argument: bool,
+    // help text to display for the option
+    pub help: &'static str,
+    // longer text to use when generating documentation/website/...
+    pub documentation: &'static str,
+}
+
+/// Function to create a clap matcher from a Configuration struct and use this to parse the
+/// command line parameters and return the processed config
+///
+/// The general flow is like this:
+/// 1. Create clap matcher from Configuration
+/// 2. Use matcher to parse command line arguments
+/// 3. If --no-config parameter was specified return parsed config
+/// 4. If --no-config is not present check environment variable STACKABLE_CONFIG_PATH
+///    if an external config file is specified
+/// 5. Parse config from file and prepend all options to the command line arguments
+/// 6. Re-parse combined arguments
+/// 7. Return parsed config
+///
+/// This effectively means that config can be either provided on the command line, or
+/// in a file that is specified via environment variable, with options from the command
+/// line taking precedence over the config file.
+///
+/// * `config` The definition of a config that the matcher will be built from
+/// * `config_file_env` Name of the environment variable to read an extra config file from
+/// * `args` The command line parameters to parse the configuration from (first element will be
+/// ignored, as this is the binary name
+pub fn get_matcher<'a>(
+    config: &dyn ConfigDescription,
+    config_file_env: &str,
+    args: Vec<OsString>,
+) -> Result<ArgMatches<'a>, Error> {
```

Comment:
I might miss something but if this can't fail (I see nothing actually returning an error) do we need to wrap it in a Result?

That said....there must be possible failures somewhere, no?

Reply:
Absolutely, I've used the absence of a logging framework as an excuse to push properly handling those back a little ;)

Reply:
Should we leave a TODO here?